code/datums/repositories/decls.dm

```diff
@@ -60,9 +62,25 @@ var/global/repository/decls/decls_repository = new
 		. = get_decls(subtypesof(decl_prototype))
 		fetched_decl_subtypes[decl_prototype] = .
 
+/repository/decls/proc/get_decl_by_id(var/id)
+	. = decls_by_id[id]
+	if(!.)
+		for(var/decl in typesof(/decl))
+			var/decl/D = GET_DECL(decl)
+			if(D.uid == id)
+				return D
+
+/decl
+	var/uid //unique identifier. If this is not hardcoded, it is dynamically generated on runtime in Initialize() below.
+
 /decl/proc/Initialize()
 	SHOULD_CALL_PARENT(TRUE)
 	SHOULD_NOT_SLEEP(TRUE)
+	if(!uid)
+		var/uid_pregen = replacetext("[type]", "/", "_")
+		uid_pregen = replacetext(uid_pregen, "_decl", "")
+		uid = "AUTOGEN[uid_pregen]"
+	decls_repository.decls_by_id[uid] = src
```

Comment:
Might be worth adding `if(decls_repository.decls_by_id[uid]) PRINT_STACK_TRACE("Duplicate /decl uid for [type]: [uid].")` before this line, though I suppose the unit test will catch that if you add one.